Add option to forward all mails send in dev env to external email.

hackerkid opened the pull request:
Fixes #7085

hackerkid commented:
@timabbott Thsee send_email_smtp function in EmailLogBackend don't have test coverage as I am mocking calls to that function in backend tests. This lost the 100% coverage of backeneds.py. What should be the strategy here?

timabbott commented:
I think we'll want to separate out `zproject/auth_backends.py `from `zproject/email_backends.py`.  You can start by moving this stuff to `email_backends.py`.

timabbott commented:
I merged the first commit.  For the second, I think it might be better to have the SMTP configuration piece done entirely in a config file, so that the thing you need to enter on the `/emails` page is just the email address to send to...

hackerkid commented:
>  For the second, I think it might be better to have the SMTP configuration piece done entirely in a config file, so that the thing you need to enter on the /emails page is just the email address to send to.

Alright. In that case should I store the email address to which all the mails sent in development environment to be forwarded and whether this feature is enabled or not in the same config file?

timabbott commented:
I think having a default email there (i.e. something to be prefilled in the UI widget to send), but I like having it show you the email explicitly.  

For the new file, I wonder whether we should just have folks put things in `zproject/dev-secrets.conf`?  We can maybe have `dev_settings.py` call `get_secret` to read items from it.  It's not clear we need a new file.

hackerkid commented:
Okay. In that case I think I can pass the value of email_address to which we should send the emails and whether this option is enabled or disabled to EmailLogBackEnd in environment variables? Does that makes sense.

timabbott commented:
I don't see why an environment variable would be needed; we'd have a Django conf option/setting that's the initial state for the text box, but the actual email to send to would be in the POST just as it is now.

hackerkid commented:
I think we are both thinking about slightly different features. I was imagining a feature that will forward all the emails sent in the dev environment to a particular email address unless it is disabled. I think you are proposing a feature in which we will generate and send the emails  to the email entered in text box when one click on a button to send them.

timabbott commented:
Yeah, I think that's the source of the confusion.  I think probably for most testing applications, the feature I had in mind makes more sense; usually you'll have a specific email that you want to verify renders properly in a real email client, not want every email generated in Zulip to get sent there...

timabbott commented:
I don't think it'll be possible to pass data around usefully using environment variables, but if passing the data around is annoying, I think it'd be fine to just have the email to forward to be set in the config file.

timabbott commented on zerver/views/email_log.py:
```diff
@@ -15,14 +21,18 @@
 ZULIP_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), '../../')
 client = Client()
 
+@csrf_exempt
```
Why does this need csrf_exempt?

hackerkid commented:
> Why does this need csrf_exempt?

I was no sending csrf_token in Post request as the form is used only in dev environment. Updated form to send csrf token in post request and removed csrf_exempt.

hackerkid commented:
@timabbott I have fixed the merge conflicts.

timabbott commented:
This looks great; I tweaked a few strings and merged this, thanks @hackerkid!  

A few follow-ups:
* I had sorta been imagining we'd use e.g. `get_secret('email_host')` for `EMAIL_HOST` and `EMAIL_HOST_USER` in `zproject/dev_settings.py`, so that you can make changes only in `dev-secrets.conf` (which has the nice property of being not tracked by Git, so something you can change once and maintain later).  What do you think of that change?  I think it'd be a lot better that way.
* I'm curious whether you think the "remember to restart the server" instructions should have actually been swapped further up in `config_error.html`.

Thanks @hackerkid!

hackerkid commented:
@timabbott Yeah. I will create a PR to make that change. I think we should also do something similiar to `GOOGLE_OAUTH2_CLIENT_ID`.

> I'm curious whether you think the "remember to restart the server" instructions should have actually been swapped further up in config_error.html.

I thought it should be at bottom. If you think the old position was correct I will bring it back.

timabbott commented:
> I thought it should be at bottom. If you think the old position was correct I will bring it back.

I'm not sure which is better, so maybe we should just stick with it as you have it :)
